### PR TITLE
FullyQualified package names

### PR DESCRIPTION
**Pull request by kumare3**
Signed-off-by: Ketan Umare <ketan.umare@gmail.com>

# TL;DR
This PR makes it possible to always get consistent package names.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [ ] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Complete description
by default the package name for a workflow or a task is relative from where the `package` or `serialize` commands are run. After this change the package name will always be relative to the root. 

The package name is deduced using hierarchical parsing of the file-system tree, to identify the root. The root is identified if the the directory does not contain a __init__.py file. For cases like flytekitplugins, it works as the module normalizer, assumes that the newly constructed path is always larger than the assumed path. If not, then the original path is returned.

Furthermore this feature can be turned of by setting the following environment variable. 

```
export FLYTE_PYTHON_PACKAGE_ROOT="."
OR
export FLYTE_PYTHON_PACKAGE_ROOT="/acutal-path/to/root/"
``` 

The default value is `FLYTE_PYTHON_PACKAGE_ROOT="auto"` which turns it on.

## Tracking Issue
https://github.com/flyteorg/flyte/issues/2241

## Follow-up issue
_NA_

### CODE REVIEW

**Review by wild-endeavor — APPROVED** [review]

**Comment by convexquad** [comment]
Just an FYI that if someone later looks at this PR - at Woven Planet we are indeed setting `export FLYTE_PYTHON_PACKAGE_ROOT="."` before serializing our workflows for compatibility with our Bazel build system (which installs Python dependencies like `flytekit` at a deeply nested path in order to provide isolation from system dependencies).
